### PR TITLE
GP Updates - July 2024

### LINKED ISSUE #26487: [INTERNAL] GP Updates
### Describe the issue

Issue to link GP update PRs to.

### Expected behavior

GP Updates merged.

### Steps to reproduce

N/A

### Additional context

_No response_

### I will provide a fix for a bug

- [X] I will provide a fix for a bug

### PR DESCRIPTION
**Pull request by jaymckinney-enavate**
This update enhances the GP migrations by:

- Migrating GP Kit Items as BC Assembly BOMs.
- Prevent migration error when Customer/Vendor phone numbers contain alpha characters. The Customer/Vendors will be migrated, but the phone/fax numbers with alpha characters will be ignored and added to the Warning Log.
- History Gen. Journal Line - View Distributions feature.
- Updated labels in settings area to specify 'migration' instead of 'upgrade' to prevent customer confusion.

Fixes #26487
Fixes [AB#543399](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/543399)

### CODE REVIEW

File: Apps/W1/DynamicsGPHistoricalData/app/src/Pages/HistReceivablesDocument.Page.al [hunk 1]
```diff
@@ -185,6 +185,34 @@ page 41006 "Hist. Receivables Document"
...
+                Caption = 'View Distributions';
+                ToolTip = 'View the G/L account distributions related to this transaction.';
+                Image = RelatedInformation;
+
+                trigger OnAction()
+                var
+                    HistGenJournalLines: Page "Hist. Gen. Journal Lines";
+                begin
+                    HistGenJournalLines.SetFilterOriginatingTrxSourceNo(Rec."Audit Code");
+                    HistGenJournalLines.RunModal();
```
Comment (nikolakukrika) [inline Apps/W1/DynamicsGPHistoricalData/app/src/Pages/HistReceivablesDocument.Page.al:210]: Minor comment here - does it need to be RunModal?
RunModal will block any other code from running, so the calling thread will wait for the dialog to close. This is needed if the page is doing the changes to the data that is shown in the main page or if we want to run some code afterwards so we need to wait. From this code it does not seem to be the case, so all calls could be changed to .Run().

If we need to wait, won't fix this comment.

File: Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al [hunk 2]
```diff
@@ -46,26 +47,38 @@ codeunit 139662 "GP Item Tests"
             Migrate(GPItem);
         until GPItem.Next() = 0;
 
-        // [THEN] A Item is created for all staging table entries
-        Assert.RecordCount(Item, GPItem.Count());
-        Assert.AreEqual(GPItem.Count(), HelperFunctions.GetNumberOfItems(), 'Wrong number of Items calculated');
+        HelperFunctions.CreatePostMigrationData();
+
+        // [THEN] An Item is created for all configured staging table entries
+        Assert.RecordCount(Item, 7);
```
Comment (nikolakukrika) [inline Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al:53]: Minor, lets not hardcode 7, could we have a variable at least close where the data is created? The ideal solution would be to calculate it from some data, if it is possible.

Comment (nikolakukrika) [inline Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al:53]: Minor, lets not hardcode 7, could we have a variable at least close where the data is created? The ideal solution would be to calculate it from some data, if it is possible.

File: Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al [hunk 3]
```diff
@@ -46,26 +47,38 @@ codeunit 139662 "GP Item Tests"
...
-            Assert.AreEqual(GPItem.StandardCost, Item."Standard Cost", 'Standard Cost not set');
-            Assert.AreEqual(GPItem.ShipWeight, Item."Net Weight", 'Net Weight not set');
-            Assert.AreEqual(GPItem.BaseUnitOfMeasure, Item."Base Unit of Measure", 'Base Unit of Measure not set');
-            Assert.AreEqual(GPItem.Description, Item.Description, 'Description not set.');
-            Assert.AreEqual(GPItem.ShortName, Item."Description 2", 'Description2 not set.');
-            Assert.AreEqual(GPItem.SearchDescription, Item."Search Description", 'Search Description not set.');
-            Assert.AreEqual(GPItem.PurchUnitOfMeasure, Item."Purch. Unit of Measure", 'Purch. Unit of Measure not set.');
-            GPItem.Next();
-        until Item.Next() = 0;
+        GPItem.Get('1 1/2\"SASH BRSH');
```
Comment (nikolakukrika) [inline Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al:57]: Similar comment here, lets introduce a variable at least if it is not possible to calculate the data

File: Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al [hunk 4]
```diff
@@ -46,26 +47,38 @@ codeunit 139662 "GP Item Tests"
...
+        Assert.AreEqual(GPItem.CurrentCost, Item."Unit Cost", 'Unit Cost not set');
+        Assert.AreEqual(GPItem.StandardCost, Item."Standard Cost", 'Standard Cost not set');
+        Assert.AreEqual(GPItem.ShipWeight, Item."Net Weight", 'Net Weight not set');
+        Assert.AreEqual(GPItem.BaseUnitOfMeasure, Item."Base Unit of Measure", 'Base Unit of Measure not set');
+        Assert.AreEqual(GPItem.Description, Item.Description, 'Description not set.');
+        Assert.AreEqual(GPItem.ShortName, Item."Description 2", 'Description2 not set.');
+        Assert.AreEqual(GPItem.SearchDescription, Item."Search Description", 'Search Description not set.');
+        Assert.AreEqual(GPItem.PurchUnitOfMeasure, Item."Purch. Unit of Measure", 'Purch. Unit of Measure not set.');
+
+        GPItem.Get('4'' STEPLADDER');
```
Comment (nikolakukrika) [inline Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al:70]: Similar comment regarding hardcoding

**Review by nikolakukrika — DISMISSED** [review]

File: Apps/W1/DynamicsGPHistoricalData/app/src/Pages/HistReceivablesDocument.Page.al (hunk 1 above)
Comment (jaymckinney-enavate) [reply Apps/W1/DynamicsGPHistoricalData/app/src/Pages/HistReceivablesDocument.Page.al:210]: Will update to .Run(), thanks.

File: Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al (hunk 2 above)
Comment (jaymckinney-enavate) [reply Apps/W1/HybridGP/test/src/GPItemTests.codeunit.al:53]: Removed, this line is not needed. The next line is what should be tested. Thanks.
Assert.AreEqual(Item.Count(), HelperFunctions.GetNumberOfItems(), 'Wrong number of Items calculated');

**Review by nikolakukrika — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: 23032.

**Pull request merged**